src/SubscriberBase.h

```diff
@@ -0,0 +1,55 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+#pragma once
+
+#include <folly/MoveWrapper.h>
+#include "src/EnableSharedFromThis.h"
+#include "src/ReactiveStreamsCompat.h"
+#include "src/mixins/ExecutorMixin.h"
+
+namespace reactivesocket {
+
+template <typename T>
+class SubscriberBaseT : public Subscriber<T>,
+                        public EnableSharedFromThisBase<SubscriberBaseT<T>>,
+                        public virtual ExecutorBase {
+  virtual void onSubscribeImpl(std::shared_ptr<Subscription> subscription) = 0;
+  virtual void onNextImpl(T payload) = 0;
+  virtual void onCompleteImpl() = 0;
+  virtual void onErrorImpl(folly::exception_wrapper ex) = 0;
+
+ public:
+  using ExecutorBase::ExecutorBase;
+
+  void onSubscribe(std::shared_ptr<Subscription> subscription) override final {
+    runInExecutor(std::bind(
+        &SubscriberBaseT::onSubscribeImpl,
+        this->shared_from_this(),
```

Comment:
I know this looks strange. But without this-> the code would not compile. The reason why is that the name resolution in the templated code is excluding dependent base (base class dependent on a template argument, in our case it is EnableSharedFromThisBase<SubscriberBaseT<T>>) from lookup:
https://gcc.gnu.org/wiki/VerboseDiagnostics#dependent_base